draft-irtf-cfrg-hpke.md

```diff
@@ -375,133 +375,126 @@ and {{encryption-context}} for details.
-
-Then, assuming that finding the discrete log of uniform random elements of `G` is
-computationally intractible, we can  implement the KEM related cryptographic dependencies
-of HPKE from the KDF and the interface to `G` as follows. The resulting KEM is called 
-`DHKEM(Group, KDF)` where `Group` denotes the Discrete Log group and `KDF` the KDF. 
+  - `sk_gen()`: returns a fresh uniform random secret key scalar
+  - `DeriveKeyPair(ikm)`: deterministically derives a key pair from `ikm`,
+    as described in {{derive-key-pair}}.
+  - `split(str, L)`: splits the byte string `str` into two substrings,
+    the first one consisting of the `L` first bytes of `str`, and the
+    second one of the remaining bytes of `str`.
+
+Then, assuming that finding the discrete log of uniformly random elements of `G` is
+computationally intractible, we can  implement the KEM-related cryptographic dependencies
+of HPKE from the KDF and the interface to `G` as follows. The resulting KEM is called
+`DHKEM(Group, KDF)` where `Group` denotes the Discrete Log group and `KDF` the KDF.
 The function parameters `pkR` and `pkS` are deserialized public keys (i.e. points in `G`).
-{{derive-key-pair}} contains the `DeriveKeyPair` function specification for DHKEMs 
+{{derive-key-pair}} contains the `DeriveKeyPair` function specification for DHKEMs
 defined in this document.
 
-* Constants used internally by DHKEM:
-  - `zero_string = 0x00 = int_to_string(0,1)` : an octet with value `0`
-  - `encap_string = 0x10 = int_to_string(8,1)` : an octet with value `8`
-  - `authencap_string = 0x11 = int_to_string(9,1)` : an octet with value `9`
-  
 * Parameters:
-  - `Npk = ptLen`: length in octets of serialized public
-  - `Nsk = 2n`: length in octets of serialized secret key
-  - `Nenc = ptLen`: length in octets of ciphertext produced by `Encap()`
-  - `Naenc = ptLen + 2n`: length in octets of ciphertext produced by `AuthEncap()` 
+  - `Npk = Np`: length in bytes of a serialized public key
+  - `Nsk = Nf`: length in bytes of a serialized secret key
+  - `Nenc = Np`: length in bytes of a KEM ciphertext produced by `Encap()`
+  - `Naenc = Np + Nf`: length in bytes of a KEM ciphertext produced by `AuthEncap()`
 
 * Algorithms:
 
 ~~~
-def ExtractAndExpand(dh, kem_context):
-  eae_prk = LabeledExtract("", "eae_prk", dh)
-  shared_secret = LabeledExpand(eae_prk, "shared_secret", kem_context, Nsecret)
-  return shared_secret
-  
+def ExtractAndExpand(ikm, label, context, L):
+  eae_prk = LabeledExtract("", "eae_prk", ikm)
+  h = LabeledExpand(eae_prk, label, context, L)
+  return h
+
 def GenerateKeyPair():
   sk = sk_gen()
   pk = sk . B
   return sk, pk
 
 def Encap(pkR):
-  skE, pkE = GenerateKeyPair()  
-  enc = point_to_string(pkE)
-  dh = skE . pkR
-  ikm = suite_id || \
-           encap_string || \
-           point_to_string(dh)  || \
-           point_to_string(pkE) || \
-           point_to_string(pkR)
-  shared_secret = ExtractAndExpand(ikm, zero_string)
```

Comment:
Using the context in the Extract step of HKDF is not recommended (see https://hyp.is/BEfR7m1AEeqBfpNgYg1Ivg/tools.ietf.org/html/rfc5869 https://tools.ietf.org/html/rfc5869#section-3.3). I would advise to split it up: `dh` as first parameter, the rest as second parameter.